internal/guest/storage/scsi/scsi.go

```diff
@@ -168,27 +217,50 @@ func Mount(
 			}
 		}
 		source = encryptedSource
-		mountType = "xfs"
 	}
 
-	for {
-		if err := unixMount(source, target, mountType, flags, data); err != nil {
-			// The `source` found by controllerLunToName can take some time
-			// before its actually available under `/dev/sd*`. Retry while we
-			// wait for `source` to show up.
-			if errors.Is(err, unix.ENOENT) || errors.Is(err, unix.ENXIO) {
-				select {
-				case <-ctx.Done():
-					log.G(ctx).Warnf("mount system call failed with %s, context timed out while retrying", err)
-					return err
-				default:
-					time.Sleep(10 * time.Millisecond)
-					continue
+	// Get the filesystem that is already on the device (if any) and use that
+	// as the mountType unless `Filesystem` was given.
+	deviceFS, err := _getDeviceFsType(source)
```

Comment:
Docs suggest that ext4 superblock is 4KiB. When checking for ext4 FS, we skip 1KiB padding and then read the superblock. Which means that we end up reading the first 5KiB. In the case of encrypted scratch, only the first 4KiB of dm-crypt device will be written to and not cause an integrity error when read from, but the remaining 1KiB can potentially cause an integrity error and we'll return on L227, because the error will be different from `ErrUnknownFilesystem`. I think we need to zero an extra block, just to be safe inside `EncryptDevce`.

Reply:
@anmaxvl The docs [here](https://ext4.wiki.kernel.org/index.php/Ext4_Disk_Layout) mention that the superblock is 1024 bytes. Are these out of date or where did you see the 4KiB number?

Reply:
Maybe it would be better to just do:

```golang
var deviceFS
if config.Encrypted {
    // set up verity/crypt
} else {
    deviceFS, err := _GetDeviceFsType(source)
    ...
}
```

So we always treat the device as containing no fs if we set up encryption.

Reply:
> @anmaxvl The docs [here](https://ext4.wiki.kernel.org/index.php/Ext4_Disk_Layout) mention that the superblock is 1024 bytes. Are these out of date or where did you see the 4KiB number?

it mentioned that the default block size is 4KiB and in the layout table mentions that SB is 1 block... maybe it's also worth special casing like @kevpar suggested.

Reply:
Added the check that Kevin suggested in.